File: bto_management/services/auth_service.py
# services/auth_service.py
from repositories.user_repository import UserRepository
from models.user import User
from utils.exceptions import AuthenticationError, OperationError, DataSaveError
from utils.helpers import validate_nric

class AuthService:
    """Handles user authentication and password management."""
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    def login(self, nric: str, password: str) -> User:
        """Authenticates a user based on NRIC and password."""
        if not validate_nric(nric):
            raise AuthenticationError("Invalid NRIC format.")

        user = self.user_repository.find_user_by_nric(nric)

        if user is None:
            raise AuthenticationError(f"NRIC '{nric}' not found.")

        # In a real system, use hashed passwords and verification
        if user.password != password:
            raise AuthenticationError("Incorrect password.")

        print(f"User {user.name} ({user.get_role().name}) logged in successfully.")
        return user

    def change_password(self, user: User, current_password: str, new_password: str):
        """Changes the password for the given user after verifying the current one."""
        if not new_password or new_password.isspace():
            raise OperationError("New password cannot be empty.")

        # Verify current password before changing
        if user.password != current_password:
             raise OperationError("Incorrect current password.")

        # Basic validation (add more rules if needed)
        if len(new_password) < 4: # Example minimum length
             raise OperationError("New password must be at least 4 characters long.")

        if new_password == user.password:
             raise OperationError("New password cannot be the same as the old password.")

        # Update the user object and save
        user.password = new_password
        try:
            self.user_repository.save_user(user)
            print(f"Password changed successfully for user {user.nric}.")
        except DataSaveError as e:
            # Revert the password change on the object if save fails?
            # For simplicity, we'll just report the error.
            # user.password = current_password # Revert if needed
            raise OperationError(f"Failed to save new password due to a data error: {e}")
        except Exception as e:
            # Catch unexpected errors during save
            raise OperationError(f"An unexpected error occurred while saving the new password: {e}")

    def get_user_role(self, user: User):
         """Safely gets the user's role."""
         try:
             return user.get_role()
         except Exception as e:
             print(f"Error getting role for user {user.nric}: {e}")
             return None # Or a default role / raise error

File: J2/Services/AuthService.java
package Services;

import Interfaces.Repositories.IUserRepository;
import Interfaces.Services.IAuthService;
import Models.User;
import Utils.NricValidator;

public class AuthService implements IAuthService {
    private final IUserRepository userRepository;

    public AuthService(IUserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public User login(String nric, String password) {
        if (nric == null || password == null) return null;
        String upperNric = nric.trim().toUpperCase();

        if (!NricValidator.isValidNric(upperNric)) {
            System.out.println("Invalid NRIC format.");
            return null;
        }
        User user = userRepository.findUserByNric(upperNric); // Use repository method
        if (user != null && user.getPassword().equals(password)) {
            return user;
        } else if (user != null) {
            System.out.println("Incorrect password.");
            return null;
        } else {
            System.out.println("User NRIC not found.");
            return null;
        }
    }

    @Override
    public boolean changePassword(User user, String oldPassword, String newPassword) {
        if (user == null || oldPassword == null || newPassword == null) return false;

        // Find the user in the repository to ensure we're working with the persisted state if necessary
        // Though the passed 'user' object should be the one to modify
        User persistedUser = userRepository.findUserByNric(user.getNric());
        if (persistedUser == null || !persistedUser.getNric().equals(user.getNric())) {
             System.err.println("Error: User mismatch during password change.");
             return false;
        }


        if (user.getPassword().equals(oldPassword)) {
            if (!newPassword.isEmpty()) {
                if (newPassword.equals(oldPassword)) {
                     System.out.println("New password cannot be the same as the old password.");
                     return false;
                }
                user.setPassword(newPassword); // Update the user object passed in

                // Save the updated user state via the repository
                Map<String, User> allUsers = userRepository.getAllUsers();
                allUsers.put(user.getNric(), user); // Update the map
                userRepository.saveUsers(allUsers); // Persist changes

                System.out.println("Password changed successfully. Please log in again.");
                return true;
            } else {
                 System.out.println("New password cannot be empty.");
                 return false;
            }
        } else {
            System.out.println("Incorrect old password.");
            return false;
        }
    }
}
